src/lib/trakeva_cache.ml

```diff
@@ -0,0 +1,195 @@
+(**************************************************************************)
+(*  Copyright 2014, Sebastien Mondet <seb@mondet.org>                     *)
+(*                                                                        *)
+(*  Licensed under the Apache License, Version 2.0 (the "License");       *)
+(*  you may not use this file except in compliance with the License.      *)
+(*  You may obtain a copy of the License at                               *)
+(*                                                                        *)
+(*      http://www.apache.org/licenses/LICENSE-2.0                        *)
+(*                                                                        *)
+(*  Unless required by applicable law or agreed to in writing, software   *)
+(*  distributed under the License is distributed on an "AS IS" BASIS,     *)
+(*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or       *)
+(*  implied.  See the License for the specific language governing         *)
+(*  permissions and limitations under the License.                        *)
+(**************************************************************************)
+
+open Nonstd
+open Pvem_lwt_unix
+open Pvem_lwt_unix.Deferred_result
+module String = StringLabels
+
+let debug = ref false
+
+let dbg fmt = ksprintf (eprintf "Trakeva_cache: %s\n%!") fmt
+    
+module Greedy_cache = struct
+
+  type t = {
+    collections: (string, (string, string) Hashtbl.t) Hashtbl.t;
+  }
+  let create () = {collections = Hashtbl.create 10}
+  let find_collection {collections} c =
+    try Some (Hashtbl.find collections c)
+    with _ -> None
+
+  let from_collection c key =
+    try Some (Hashtbl.find c key)
+    with _ -> None
+    
+  let new_collection t collection =
+    let col = Hashtbl.create 42 in
+    Hashtbl.add t.collections collection col;
+    col
+
+  let add_to_collection t col ~key ~value =
+    Hashtbl.add col key value
+
+  let fold_collection col ~init ~f =
+    let r = ref init in
+    Hashtbl.iter (fun key value ->
+        if !debug then dbg "fold_collection %s, %s" key value;
+        r := f !r ~key ~value) col;
+    !r
+
+
+
+end
+
+
+module Add (KV_DB: Trakeva.KEY_VALUE_STORE): Trakeva.KEY_VALUE_STORE = struct
+
+  let default_collection_name = "trakeva-greedy-cache-default-collection"
+    
+  type t = {
+    original: KV_DB.t;
+    cache: Greedy_cache.t;
+    mutex: Lwt_mutex.t;
+  }
+  let load s =
+    KV_DB.load s
+    >>= fun original ->
+    return {original; cache = Greedy_cache.create (); mutex = Lwt_mutex.create ()}
+
+  let close {original; cache} =
+    KV_DB.close original
```

Comment:
The `cache` in `{original; cache}` is superfluous. Were you going to do something with it?